### PR TITLE
Intrepid2: Support Orientations with Structured Integration

### PR DESCRIPTION
**Pull request by CamelliaDPG**
@trilinos/intrepid2

## Motivation
Structured integration to date (i.e., support for sum factorization in tensor-product bases, with the hope to extend to support for matrix-free computations) has not explicitly supported application of element orientations; assembly has been performed locally, with global assembly entirely left to the user.  (Existing support for orientations in Intrepid2 has assumed application of orientations to basis values, effectively erasing the structural information required for sum factorized assembly.)

This PR addresses this by adding support for application of orientations to an assembled local matrix.  The workflow required is as follows:
1. Record orientations into a (C) View of orientations, where C is the cell index.  (See `CellGeometry::orientations()`.)
2. Perform local assembly into a (C,F1,F2) View, where F1 is the field index for the left basis, and F2 is the field index for the right basis.  (See `IntegrationTools::integrate()`.)
3. Call the new `OrientationTools::modifyMatrixByOrientation()`, passing in the local, unoriented (C,F1,F2) matrix, the orientations, and the left and right basis.  This will store into an oriented (C,F1,F2) matrix.

It's worth noting that in the local assembly, one should use the local/unoriented basis for everything.  If doing a nonlinear problem, then the basis coefficients used there should be expressed not in terms of the global solution (oriented basis), but in terms of the local basis.

To transform global coefficients (on an oriented basis) into local coefficients (on an unoriented basis) is to apply the transpose of the orientation matrix that took you from local basis values to global basis values.  This can be accomplished using the new `OrientationTools::modifyBasisByOrientationTranspose()`.

This PR also adds the following:
- Made all the tests comparing standard to structured assembly in test group `StructuredIntegration` use orientations, so that the new orientation support is exercised in all those tests
- Added a more general, asymmetric mechanism for performing both standard and structured assembly of one basis function integrated against another, in `assembly-examples/StandardAssembly.hpp` and `assembly-examples/StructuredAssembly.hpp`, respectively.
- Added a host of asymmetric tests in the test group `StructuredIntegration`, against 1D, 2D, and 3D examples and against all function spaces, and a few polynomial order combinations (p=1 and p=2).
- Split `StructuredIntegrationTests.cpp` into seven files (five `.cpp` files, and two `.hpp` files) for cleaner organization and faster compilation.
- Resolved a couple bugs in the kernels for `IntegrationTools::integrate()` revealed by the new tests.
- In Data:
    - in `storeInPlaceCombination()`: fixed the ArgExtractor for a case when B (not B_underlying) is used; was incorrectly using the underlying ArgExtractor.
    - added a constructor with run-time logical rank (but compile-time View rank)
    - fixed the `setUnderlyingViewN()` methods to be non-const
    - fixed `shallowCopy()` to properly set the rank of its returned Data object
 -  Added `DataTools`:
    - Utility methods for manipulating `Intrepid2::Data` objects
    - Initial method is multiplyByCPWeights() which allocates, fills, and returns a Data container that is the result of a (C,P,D,D) container weighted by values from a (C,P) container.
 -  In `TransformedBasisValues`, added `multiplyByPointwiseWeights()` method to modify the transformed basis values by user-supplied (C,P) weights
 - In `CellTools`, added method `setJacobianDividedByDet()`.

## Testing
Adds a bunch more tests, including tests against the new orientation support:
- `OrientationTools::modifyBasisByOrientationTranspose()` is tested in the context of orientation tests in `test_orientation_HEX_newBasis.hpp`, where we test directly that a global, oriented solution sum agrees with a local, unoriented sum.
- `OrientationTools::modifyMatrixByOrientation()` is tested in the context of our structured integration verification tests; the tests compare values computed by structured integration with those computed by standard integration.  Standard integration uses the pre-existing `OrientationTools::modifyBasisByOrientation()` to account for orientations.
- Unit tests in `TransformedBasisValuesTests.cpp` cover the `multiplyByPointwiseWeights()` method.

### CODE REVIEW

File: packages/intrepid2/assembly-examples/StructuredAssembly.hpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  Data<PointScalar,DeviceType> jacobian       = geometry.allocateJacobianData(tensorCubaturePoints, 0, worksetSize);
+  Data<PointScalar,DeviceType> jacobianDividedByJacobianDet = geometry.allocateJacobianData(tensorCubaturePoints, 0, worksetSize); // jacobianDividedByJacobianDet has same underlying structure as jacobian
+  Data<PointScalar,DeviceType> jacobianDet    = CellTools<DeviceType>::allocateJacobianDet(jacobian);
+  Data<PointScalar,DeviceType> jacobianDetInv = CellTools<DeviceType>::allocateJacobianDet(jacobian);
+  Data<PointScalar,DeviceType> jacobianInv    = CellTools<DeviceType>::allocateJacobianInv(jacobian);
+  TensorData<PointScalar,DeviceType> cellMeasures = geometry.allocateCellMeasure(jacobianDet, tensorCubatureWeights);
+  
+  // lazily-evaluated transformed basis values (temporary to allow integralData allocation)
+  auto transformedBasis1ValuesTemp = transform(basis1Values, fs1, op1, jacobian, jacobianDet, jacobianInv, jacobianDetInv, jacobianDividedByJacobianDet);
+  auto transformedBasis2ValuesTemp = transform(basis2Values, fs2, op2, jacobian, jacobianDet, jacobianInv, jacobianDetInv, jacobianDividedByJacobianDet);
```
Comment (GrahamBenHarper) [inline packages/intrepid2/assembly-examples/StructuredAssembly.hpp:162]: Being able to do structured integration by passing the differential operators `op1` and `op2` was something I didn't realize would be possible, but that's really useful for the matrix-free work I'm doing!

I think these sort of interfaces will also be useful for integrating structured and matrix-free capabilities with other packages like Panzer.

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationDef.hpp [hunk 2]
```diff
@@ -97,8 +97,8 @@ namespace Intrepid2 {
     for(ordinal_type i=0;i<numVerts-1;++i)
       for(ordinal_type j=i+1;j<numVerts;++j)
         INTREPID2_TEST_FOR_ABORT( ( subCellVerts[i] == subCellVerts[j] ), 
-                        ">>> ERROR (Intrepid::Orientation::getOrientation): " \
-                        "Invalid subCellVerts, same vertex ids are repeated");
+                                 ">>> ERROR (Intrepid::Orientation::getOrientation): " \
+                                 "Invalid subCellVerts, same vertex ids are repeated");
```
Comment (GrahamBenHarper) [inline packages/intrepid2/src/Orientation/Intrepid2_OrientationDef.hpp:101]: I saw you correct this to "some vertex ids" in other places

**Review by GrahamBenHarper — DISMISSED** [review]
I just have some general questions or comments. Overall it looks like it "integrates" well with the existing structured integration features :) Thanks for putting this together @CamelliaDPG!

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationDef.hpp (hunk 2 above)
Comment (CamelliaDPG) [reply packages/intrepid2/src/Orientation/Intrepid2_OrientationDef.hpp]: Ah, thanks for catching this.  I think that correction got clobbered when I merged latest develop into this branch.

File: packages/intrepid2/src/Shared/Intrepid2_DataTools.hpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+class DataTools
+{
+public:
+  //! Allocates and fills Data object of logical shape (C,P,D,D) corresponding to the pointwise product of an object of shape (C,P,D,D) with one of shape (C,P).
+  //! Will also work for any "matrix" data and scalar data which differ in rank by 2, but otherwise share the same shape.  E.g., (C,F,P,D1,D2) matrices could be multiplied by (C,F,P) scalars.
+  //! \param matrixDataIn [in] - the (C,P,D,D) container.
+  //! \param scalarDataIn [in] - the (C,P) container.
+  //! \return
+  template<class Scalar, class DeviceType>
+  static Data<Scalar,DeviceType> multiplyByCPWeights(const Data<Scalar,DeviceType> &matrixDataIn, const Data<Scalar,DeviceType> &scalarDataIn)
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Shared/Intrepid2_DataTools.hpp:64]: Yes, that's actually the intended use case, although I don't necessarily expect most users to call this method directly.  They should instead compute their `TransformedBasisValues` using FunctionSpaceTools, and then do:

```
transformedBasisValues.multiplyByPointwiseWeights(cpWeights);
```

They can then pass these data-weighted basis values to `integrate()` as usual.

**Review by mperego — APPROVED** [review]
Hi @CamelliaDPG, this looks great to me. Thanks also for adding extra tests/examples, in particular I think that the structured vs standard examples are very useful.

File: packages/intrepid2/assembly-examples/StandardAssembly.hpp [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    
+    FunctionSpaceTools::computeCellMeasure(cellMeasures, jacobianDeterminant, cubatureWeights);
+    ExecutionSpace().fence();
+    jacobianAndCellMeasureTimer->stop();
+    
+    // because structured integration performs transformations within integrate(), to get a fairer comparison here we include the transformation calls.
+    fstIntegrateCall->start();
+    transform(unorientedTransformedValues1, basis1Values, fs1, op1, jacobian, jacobianDeterminant, jacobianInverse);
+    transform(unorientedTransformedValues2, basis2Values, fs2, op2, jacobian, jacobianDeterminant, jacobianInverse);
+    OrientationTools<DeviceType>::modifyBasisByOrientation(transformedValues1, unorientedTransformedValues1, orientationsWorkset, basis1.get());
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/assembly-examples/StandardAssembly.hpp]: Sounds good.  I've made that change.

File: packages/intrepid2/src/Cell/Intrepid2_CellGeometry.hpp [hunk 5]
```diff
@@ -267,6 +267,12 @@ namespace Intrepid2
     //! Returns the orientations for all cells.  Calls initializeOrientations() if it has not previously been called.
     Data<Orientation,DeviceType> getOrientations();
     
+    //! \brief Fills the provided container with the orientations for the specified cell range.  Calls getOrientations() and copies the orientations from the Data container into the ScalarView container.
+    //! \param [out] orientationsView - the container that will be filled.
+    //! \param [in] startCell - the first cell ordinal whose orientation will be copied.
+    //! \param [in] endCell - the first cell ordinal whose orientation will not be copied; use -1 to indicate that orientations for all cells from startCell on should be copied.
+    void orientations(ScalarView<Orientation,DeviceType> orientationsView, const int &startCell, const int &endCell);
```
[the remark this answers is not in the document]
Comment (CamelliaDPG) [reply packages/intrepid2/src/Cell/Intrepid2_CellGeometry.hpp]: Good idea.  I've made that change.

**Comment by CamelliaDPG** [comment]
Set WIP because I'm pushing some things that I'm manually testing under CUDA; don't yet believe that I've resolved the CUDA issue (it affects only release build, not debug; something subtle is happening).

**Comment by CamelliaDPG** [comment]
I believe the CUDA failures were due to a linker bug.  I have worked around these by replacing some lambdas with functors.  Along the way I also added some ETI for Intrepid2_Data that should help with build times (which have been quite long on Weaver especially).

Anyway, removing WIP to let the auto-tester do its thing.

**Review by mperego — APPROVED** [review]
Looks good to me. It's great to have this capability! Thanks.

**Pull request merged**